folio-spring-system-user/src/main/java/org/folio/spring/service/SystemUserService.java

```diff
@@ -0,0 +1,146 @@
+
+  private final ExecutionContextBuilder contextBuilder;
+  private final SystemUserProperties systemUserProperties;
+  private final FolioEnvironment environment;
+  private final AuthnClient authnClient;
+  private final PrepareSystemUserService prepareUserService;
+  private Cache<String, SystemUser> systemUserCache;
+
+  /**
+   * Get authenticate system user.
+   * Get from cache if present (or getSystemUser otherwise). Call login expiry endpoint in case
+   * access token expired. Call login endpoint in case login-expiry endpoint returns null or it
+   * doesn't exist
+   *
+   * @param tenantId The tenant name
+   * @return {@link SystemUser} with token value
+   */
+  public SystemUser getAuthedSystemUser(String tenantId) {
+    if (systemUserCache == null) {
+      return getSystemUser(tenantId);
+    }
+
+    var user = systemUserCache.get(tenantId, this::getSystemUser);
+    var userToken = user.token();
+    var now = Instant.now();
+    if (userToken.accessTokenExpiration().isAfter(now)) {
+      return user;
+    }
+
+    systemUserCache.invalidate(tenantId);
+    user = getSystemUser(tenantId);
+    systemUserCache.put(tenantId, user);
+
+    return user;
+  }
+
+  /**
+   * Authenticate system user and return token value.
+   *
+   * @param user {@link SystemUser} to log with
+   * @return token value
+   */
+  public UserToken authSystemUser(SystemUser user) {
+    var token = getTokenWithExpiry(user);
+    if (token == null) {
+      token  = getTokenLegacy(user);
+    }
+    return token;
+  }
+
+  @Autowired(required = false)
+  public void setSystemUserCache(Cache<String, SystemUser> systemUserCache) {
+    this.systemUserCache = systemUserCache;
+  }
+
+  private SystemUser getSystemUser(String tenantId) {
+    log.info("Attempting to issue token for system user [tenantId={}]", tenantId);
+    var systemUser = SystemUser.builder()
+        .tenantId(tenantId)
+        .username(systemUserProperties.username())
+        .okapiUrl(environment.getOkapiUrl())
+        .build();
+
+    // create context for authentication
+    try (var fex = new FolioExecutionContextSetter(contextBuilder.forSystemUser(systemUser))) {
+      var token = authSystemUser(systemUser);
+      systemUser = systemUser.withToken(token);
+      log.info("Token for system user has been issued [tenantId={}]", tenantId);
+    }
+    // create context for user with token for getting user id
+    try (var fex = new FolioExecutionContextSetter(contextBuilder.forSystemUser(systemUser))) {
+      var userId = prepareUserService.getFolioUser(systemUserProperties.username())
+          .map(UsersClient.User::id).orElse(null);
+      return systemUser.withUserId(userId);
+    }
+  }
+
+  private UserToken getTokenLegacy(SystemUser user) {
+    var response =
+        authnClient.login(new UserCredentials(user.username(), systemUserProperties.password()));
+
+    if (isNull(response) || response.getStatusCode() == HttpStatusCode.valueOf(404)) {
```

Comment:
Wouldn't be better to throw an exception from authnClient.login in case of Http 404
In that case I assume it won't be needed to test for null and verify status code. Feign client allows to throw exceptions, like `NotFoundException`, `BadRequestException`, instead of analyzing Http status code. In this case, exception handling will be more natural.

The above is true for other places where Http status code is verified

Reply:
@viacheslavkol I don't see how throwing and handling `NotFoundException` helps us here. We still have to try and catch this exception. Checking the response code is the same difference. `404` is expected if the new API isn't available.

Reply:
I this particular case, yes you'd need to catch an exception and return null. But in general handling an exception is a more natural way of dealing with exceptional cases in Java. Once a feign client throws well defined exceptions instead of returning a holder object (`ResponseEntity`) with result data or error the code becomes cleaner, in particular, I think, you won't need to test for null.

In this case my suggestion is a recommendation rather, not a request

Reply:
@dmtkachenko  changes are done in method : authSystemUser(SystemUser user). Please check and review